test/unit/api.test.js

```diff
@@ -0,0 +1,301 @@
+    }
+  });
+
+  afterEach(() => {
+    sinon.restore();
+  });
+
+  for (const {
+    apiName,
+    instance,
+    method,
+    possibleCallbackPositions,
+    makeStub
+  } of generateTests()) {
+    expect(instance, apiName).to.have.property(method).that.is.a('function');
+    const functionLength = instance[method].length;
+
+    describe(`${apiName}()`, () => {
+      const resolveSuite = [];
+      const rejectsSuite = [];
+      // Use the callback positions manually defined, or use a default of [1] / [1,2] depending on function length
+      const callbackPositions =
+        possibleCallbackPositions != null
+          ? possibleCallbackPositions
+          : functionLength < 2
+          ? [1]
+          : [1, 2];
+
+      for (const argumentDecrement of callbackPositions) {
+        // For each callback position, we construct an array of arguments that contains the callback
+        // at the specified position.  We can then test various scenarios (success callback, success promise, error callback and error promise)
+        // for that given overload.
+
+        // Make an array that is the same length as the function under test
+        const args = Array.from({ length: functionLength }, (_, i) => i);
+        const callbackPosition = functionLength - argumentDecrement;
+        // Place the callback at the position we want to see it in
+        args[callbackPosition] = 'callback';
+        // truncate the array
+        args.length = functionLength - (argumentDecrement - 1);
+        const argsPassedToDriver = args.slice(0, callbackPosition);
+
+        resolveSuite.push(() =>
+          describe(`and ${apiName} is called  with (${args.join(', ')})`, () => {
+            let superPromise;
+            let actualReturnValue;
+            let callback;
+            let stubbedMethod;
+
+            before('setup success stub for callback case', function () {
+              superPromise = Promise.resolve({ message: 'success!' });
+              stubbedMethod = makeStub(superPromise);
+              callback = sinon.spy();
+              args[callbackPosition] = callback;
+              actualReturnValue = instance[method](...args);
+            });
+
+            beforeEach(async function () {
+              await superPromise.catch(error => error);
+            });
+
+            it('should return void', () => expect(actualReturnValue).to.be.undefined);
+
+            it('should call the callback with undefined error and successful result', () => {
+              expect(callback).to.have.been.calledOnce;
+              const expectedArgs = callback.args[0];
+              expect(expectedArgs).to.have.property('0', undefined);
+              expect(expectedArgs).to.have.nested.property('[1].message', 'success!');
+            });
+
+            it(`should pass only (${argsPassedToDriver.join(', ')}) to the driver api`, () => {
+              expect(stubbedMethod).to.have.been.calledOnceWith(...argsPassedToDriver);
```

Comment:
calledOnceWith allows me to omit the `undefined`. Generating the argsPassedToDriver doesn't exactly have a pattern since sometimes there's options sometimes there's other optional args. testing that each possible argument up to the callback is passed to the underlying function kinda covers this. lmk though if you're seeing something I'm not about how we could fill in the `undefined`s

Reply:
which undefined? isn't the point of this test to make sure that "only these arguments are passed to the driver API"? (per the name of the test case)

Reply:
oh I see what you mean - I don't feel strongly enough about this; resolving